include/shark/Algorithms/Trainers/CARTcommon.h

```diff
@@ -0,0 +1,156 @@
+//===========================================================================
+/*!
+ *
+ *
+ * \brief       General functions for Tree modeling.
+ *
+ *
+ *
+ * \author      K. N. Hansen, J. Kremer, J. Wrigley
+ * \date        2011-2016
+ *
+ *
+ * \par Copyright 1995-2015 Shark Development Team
+ *
+ * <BR><HR>
+ * This file is part of Shark.
+ * <http://image.diku.dk/shark/>
+ *
+ * Shark is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Shark is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Shark.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+//===========================================================================
+#ifndef SHARK_ALGORITHMS_TRAINERS_CARTCOMMON_H
+#define SHARK_ALGORITHMS_TRAINERS_CARTCOMMON_H
+
+#include <shark/Models/Trees/General.h>
+#include <shark/Data/DataView.h>
+#include <vector>
+#include <unordered_map>
+#include <utility>
+namespace shark {
+namespace detail {
+namespace cart {
+/// ClassVector
+using ClassVector = UIntVector;
+
+
+struct Attribute {
+	double value;
+	std::size_t id;
+
+	inline bool operator<(Attribute const &v) const {
+		return value < v.value;
+	}
+};
+
+
+class Index {
```

Comment:
Honestly I was struggling a little with the naming of this class. (Also with `AttributeTable`(`s`) I feel there are more sensical names out there, I just couldn't think of them. Suggestions are welcome.)

`Index` groups the functionality of the previous
`createAttributeTables()`, and
`splitAttributeTables()`
functions. Basically an `Index` indexes the dataset by attribute, providing fast sorted-order lookup of each attribute. The use case is, that we need to scan attributes in order of size when looking for a split.

How about this(below), would this be an adequate description?
```diff
+/**
+ * An index of a dataset providing fast sorted-order lookup of each attribute.
+ */
 class Index {
```

Perhaps a better name for the class would be `SortedIndex`?